proposals/0003-health-checks.md

```diff
@@ -0,0 +1,103 @@
+# Service Healthchecks
+
+* Proposal: [MP-0003](0003-health-checks.md)
+* Authors: [Clement Escoffier](https://github.com/cescoffier), [Heiko Rupp](https://github.com/pilhuhn), [Heiko Braun] (https://github.com/heiko-braun)
+
+* Status: **Awaiting review**
+
+*During the review process, add the following fields as needed:*
+
+* Decision Notes: [Discussion thread topic covering the  Rationale](https://groups.google.com/forum/#!forum/microprofile), [Discussion thread topic with additional Commentary](https://groups.google.com/forum/#!forum/microprofile)
+
+## Introduction
+
+This document describes a protocol (wireformat, semantics and possible forms of interactions) between system components that need to determine the “liveliness” of computing nodes in a bigger architecture.
+
+
+Mailinglist thread: [Discussion thread topic for that proposal](https://groups.google.com/forum/#!topic/microprofile/jIZAKiu76ys)
+
+## Motivation
+
+Health checks are used to probe the state of a computing node from another other machine (i.e. kubernetes service controller) with the primary target being cloud infrastructure environments where automated processes maintain the state of computing nodes. In this scenario health checks are used to determine if a computing node needs to be discarded (terminated, shutdown) and eventually replaced by another (healthy) instance.
```

Comment:
@ebullient agree. Applications should compensate for the failure of downstream application and possibly handle using circuit-breaker pattern etc however there are some cases where such patterns are not easy to use e.g. DB layer. In case of DB connectivity failure, application should let LB (load-balancer) or clients know that application is either in limited mode or not available (503?).

could we provide a way to configure the dependencies of a given microservice so that a regular healt-check call could check health of the dependencies (by calling dependent service's health check)? Once we pulled individual health-checks then we provide a config/strategy to merge those responses into the application's own health-check.
For most of the use cases, a simple config should be sufficient however developers can further customize the health-check logic.